cpp/python/src/worker.cpp

```diff
@@ -71,6 +71,21 @@ void Worker::populateFuturesPool()
   }
 }
 
+void Worker::clearFuturesPool()
+{
+  if (_enableFuture) {
+    ucxx_trace_req("ucxx::python::Worker::%s, Worker: %p, populateFuturesPool: %p",
+                   __func__,
+                   this,
+                   shared_from_this().get());
+    std::lock_guard<std::mutex> lock(_futuresPoolMutex);
+    PyGILState_STATE state = PyGILState_Ensure();
+    decltype(_futuresPool) newFuturesPool;
```

Comment:
*note*: I guess this also applies to `populateFuturesPool` as well.

This function takes two locks, so we need to think about lock ordering. Is there a scenario where a thread holding the GIL might be attempting to call `getFuture` (which will need to acquire `_futuresPoolMutex`), while a different thread not holding the GIL has called `clearFuturesPool` and therefore holds the lock => deadlock?

Or is it the case that only one python thread is allowed to talk to these worker functions?

Reply:
Currently yes, only one Python thread is supported (aka, tested/used). However, this is a very much valid concern as we have almost everything in place for multi-thread Python. I've filed https://github.com/rapidsai/ucxx/issues/325 to track that and fix it in a follow-up PR.